src/rmp/src/Restructure.cpp

```diff
@@ -0,0 +1,555 @@
+             countConsts(block_));
+
+  Blif blif_(openroad_, locell_, loport_, hicell_, hiport_);
+  blif_.setReplaceableInstances(path_insts_);
+  blif_.writeBlif(input_blif_file_name_.c_str());
+  debugPrint(
+      logger_, RMP, "remap", 1, "Writing blif file {}", input_blif_file_name_);
+  files_to_remove.emplace_back(input_blif_file_name_);
+
+  // abc optimization
+  bool area_mode = opt_mode_ <= Mode::AREA_3;
+  std::string abc_script_file = "ord_abc_script.tcl";
+  std::string best_blif;
+  int best_inst_count = std::numeric_limits<int>::max();
+  for (int opt_mode = static_cast<int>(area_mode) ? static_cast<int>(Mode::AREA_1) : static_cast<int>(Mode::DELAY_1);
+       opt_mode <= (static_cast<int>(area_mode) ? static_cast<int>(Mode::AREA_3) : static_cast<int>(Mode::DELAY_4));
+       opt_mode += 1) {
+    opt_mode_ = (Mode) opt_mode;
+    output_blif_file_name_ = std::string(block_->getConstName())
+                             + std::to_string(opt_mode) + "_crit_path_out.blif";
+    debugPrint(logger_,
+               RMP,
+               "remap",
+               1,
+               "Writing ABC script file {}",
+               abc_script_file);
+    if (writeAbcScript(abc_script_file)) {
+      std::string abc_command = std::string("yosys-abc < ") + abc_script_file;
+      if (logfile_ != "")
+        abc_command = abc_command + " > " + logfile_;
+      std::system(abc_command.c_str());
+    }
+    int num_instances = 0;
+    bool status
+        = blif_.inspectBlif(output_blif_file_name_.c_str(), num_instances);
+    logger_->report(
+        "Optimized to {} instance in iter {}", num_instances, opt_mode);
+    if (status && num_instances < best_inst_count) {
+      best_inst_count = num_instances;
+      best_blif = output_blif_file_name_;
+    }
+    files_to_remove.emplace_back(output_blif_file_name_);
+  }
+  files_to_remove.emplace_back(abc_script_file);
+  if (best_inst_count < std::numeric_limits<int>::max()) {
+    // read back netlist
+    debugPrint(logger_, RMP, "remap", 1, "Reading blif file {}", best_blif);
+    blif_.readBlif(best_blif.c_str(), block_);
+    debugPrint(logger_,
+               utl::RMP,
+               "remap",
+               1,
+               "Number constants after restructure {}",
+               countConsts(block_));
+  }
+
+  for (auto file_to_remove : files_to_remove) {
+    if (!logger_->debugCheck(RMP, "remap", 16))
+      std::remove(file_to_remove.c_str());
+  }
+}
+
+void Restructure::postABC(float worst_slack)
+{
+  rsz::Resizer* resizer = openroad_->getResizer();
+  bool area_mode = opt_mode_ <= Mode::AREA_3;
+  if (!area_mode) {
+    // Recompute timing
+    resizer->estimateWireParasitics();
+    open_sta_->findRequireds();
+    float new_slack;
+    sta::Vertex* worst_vertex;
+    open_sta_->worstSlack(sta::MinMax::max(), new_slack, worst_vertex);
+    if (!area_mode && new_slack < worst_slack) {
+      // failed, revert
+
+    } else {
+      // slack improved, accept
```

Comment:
tbd??

Reply:
Yes, we have not yet worked on timing aspect of re-structuring. I can remove this till we actually get to it.